docs/proposals/build-execution-using-build-run.md

```diff
@@ -0,0 +1,76 @@
+---
+title: build-execution-using-build-run
+authors:
+  - "@sbose78"
+  - "@zhangtbj"
+status: design
+---
+
+# Build execution using the BuildRun API
+
+
+## About
+
+A `BuildRun` is an immutable CR/object that represents a single execution of the `Build`.
+
+The simplest `BuildRun` looks like this:
+```
+apiVersion: build.dev/v1alpha1
+kind: BuildRun
+metadata:
+  generateName: account-service-build-
+spec:
+  buildRef:
+    name: kaniko-golang-build
+```
+
+The `BuildRun` API also provides a way to override/specify  execution time information. 
+
+As an example, a `Build` configuration should not have to accurately specify the resource requirements - the information on resource requirements is valuable at build execution time which varies from environment to environment.
+
+## Defining a BuildRun
+
+The `BuildRun`, apart from triggering a build execution, also provides the user an opportunity to specify/override a subset of Build configuration which are likely to change at build execution time.
+
+### What is deterministic
+
+#### What are we building
+* The source code that's being built into an image.
+* The application binary that's being packaged into an image.
+
+The above information wraps the integrity of the code
+that we are building into an image.
+
+#### How are we building
+* The `BuildStrategy` being used to build the image.
+* Inputs associated with the `BuildStrategy` : Dockerfile, builder image, etc.
+* The runtime base image being used to build a lean image.
+
+The above information wraps the know-how that drives the process of converting source code into an image.
+
+### What is non-deterministic
+
+* The Image being pushed to
+    * Repository
+    * Tag
+* The Service Account used to execute the build.
+* Execution-time resource requiements.
```

Comment:
Yeah, 
Git Repo is "know-what"
Strategy is "know-how".

Good point, things like  "resource limits" may in some cases be important especially where the Build CR author wants to specify what are the mandatory minimums.

Reply:
I think the tag is at `know-know` level. because it is not included in 
 WHAT is source code or build config. It is the outcome, which the end user decide how I to tag my output image (Not what will be built).
- So I think Git repo and ImageURL are "know-what", they are immutable
- Image Tag is "know-how", Git revision should be the "know-how"

"know-what" or "know-how" is the basic rule, but I also think we also need to think about is it good for user to use our build.

Let me give two examples:
1. `Build v1` In OpenShift
2. In another build project `kpack`

### OpenShift build v1
Although, in UI, the end user cannot modify the build during trigger a new build. but you can see, ALL gitsource and output are in both BuildConfig and Build:
```
kind: BuildConfig
apiVersion: build.openshift.io/v1
...
spec:
  output:
    to:
      kind: ImageStreamTag
      name: 'cf-sample-app-go-git:latest'
  ...
  source:
    type: Git
    git:
      uri: 'https://github.com/zhangtbj/cf-sample-app-go.git'
    contextDir: /
  ...


kind: Build
apiVersion: build.openshift.io/v1
...
spec:
  output:
    to:
      kind: ImageStreamTag
      name: 'cf-sample-app-go-git:latest'
    pushSecret:
      name: builder-dockercfg-pjxkz
  source:
    type: Git
    git:
      uri: 'https://github.com/zhangtbj/cf-sample-app-go.git'
    contextDir: /
  revision:
    type: Git
    git:
      commit: 8cf9569a8ff5c6cee70137476234238a4609f300
      author:
        name: Jordan
        email: zhangtbj@cn.ibm.com
      committer:
        name: GitHub
        email: noreply@github.com
      message: Update main.go
  ...
```

### In another build project kpack
The user can define them for the `image` CRD (Which like our Build): https://github.com/pivotal/kpack/blob/master/samples/image_from_git.yaml#L6-L14
```
apiVersion: build.pivotal.io/v1alpha1
kind: Image
metadata:
  name: sample-1
spec:
  tag: sample/image-from-git
  builder:
    kind: ClusterBuilder
    name: cluster-sample-builder
  serviceAccount: service-account
  source:
    git:
      url: https://github.com/buildpack/sample-java-app.git
      revision: master
```
The user can also define the Tag and Source in the `build` (Like our BuildRun):
https://github.com/pivotal/kpack/blob/master/samples/build.yaml
```
apiVersion: build.pivotal.io/v1alpha1
kind: Build
metadata:
  name: sample
spec:
  tags:
  - sample/image
  serviceAccount: serviceaccount
  builder:
    image: gcr.io/paketo-buildpacks/builder:base
  source:
    git:
      url: https://github.com/buildpack/sample-java-app.git
      revision: master
```

I think it is more convenient for end user

Reply:
I see. I feel, git revision should therefore be specify-able in the BuildRun too. Is that what you are suggesting ?

As an example, in future, when we support Build(Run) triggering from Github webhooks, we are definitely going to have a new Git revision being used every BuildRun. Therefore, the 'revision' is, in a way non-deterministic at the time of `Build` CR creation ?

Reply:
yes, I think so :)

- The image tag is related to the image tag policy
- The git revision is related to the auto-build trigger policy

For auto-build trigger policy, we may also need to consider other ones, such as:
- Trigger by builder image update (Such as buildpacks version, for example nodejs runtime is updated)
- Trigger by base image update (fix the vulnerability issue)
- Trigger by buildstrategy update (maybe some fixes)

But these three are low priority for now, I think. For a developer, I think they care more about the gir revistion trigger rebuild. For an admin, they may care about the vul issue.

Reply:
Should we distinguish between:
* git branch
* git revision/commit 
https://github.com/redhat-developer/build/issues/223

Option 1:
* In which scenario is it OK to override the git repo + branch + revision in the BuildRun ? How does Build validate that a repo is a fork of an existing GitSource in a Build's GitSource ? The build operation doesn't care, but this feels like a difficult question because: Repo & Branch is core to the integrity of the build.  

Option 2:
* Should we at all override repo+branch given that repo & branch feels like they are core to the integrity of the build ? Maybe if the repo & branch changes, it should be a new `Build` altogether?

Option 3:
Postpone the discussion/decision since we haven't designed build triggers yet :) For now, I'll note this down in the proposal as "unresolved" and we should have a separate discussion.

Reply:
Note, build v1 didn't really support build triggers from a forked repo's branches, I remember.

Reply:
Added a note on git revision being 'not necessarily deterministic'.
https://github.com/redhat-developer/build/pull/218/commits/9e198a8a096cc59ba310ddc48aaa6e57a9440d64

Reply:
we don't need to distinguish branch/revision/commit

Fork branch is a different case which the git url is totally changed.

But I agree with the git revision is the second phase design, we don't need to support it now :)

And how about the image tag? Do we have the agreement?

Reply:
( comments on Option 1): I think a fork should be a new Build, overriding `repo` in the `BuildRun` with a fork breaks the integrity of the `Build`. 

(comments on Option 2): The repo definitely NOT, basically because of my comments on Option 1. The `branch` similar to the `tag`, depends on what a modification of a `Build` can do for you(_triggers_) and if this missing design prove to be sufficient(e.g.  _"I don't need to override them on the BuildRun"_), then yes we should go for Option 3.

(comments in Option 3): @sbose78 correct me If ´m wrong, but I understood `triggers` as spawning a new `BuildRun` automatically. If this is the case, I see the following:
- we don't have this implementation yet, but we have #82 for discussion. Maybe #82 should convert first on a proposal(similar to this one), then the PR.
- I think `triggers` decision could help us on making the right decision on what can be overriden in a BuildRun and what cannot. For example(not an statement), if with `triggers`  we agree that for the first and any subsequent modification of a `Build` I will always get a new `BuildRun`(pls keeping the manual BuildRun), then probably there is no need to override `branch` and `tag` in the BuildRun, just brainstorming, it depends on what can we achieve with triggers.

Reply:
The scope of #82 itself is very specific - behave like the era where we had only Build API and no BuildRun API :) . I wonder if a proposal for that might be an overkill, 

However, I would like to have us write a proposal about "Building CLI/UI tooling and experiences around Build v2". The use case of #82 would a part of it. The proposal/guidance on 'building tooling around Build v2' will help us iron out the expectations on the different scenarios it would be used in.
* Coligo UI
* Coligo CLI
* oc new-app ?
* odo 
* Within a CI pipeline

Reply:
> For example(not an statement), if with triggers we agree that for the first and any subsequent modification of a Build I will always get a new BuildRun(plus keeping the manual BuildRun), then probably there is no need to override branch and tag in the BuildRun

Yeah, "always build from 'develop' branch" kinda thing. We should definitely fork a separate discussion on what we should support and what we shouldn't.

Reply:
@sbose78  I think a proposal around "Building CLI/UI tooling and experiences around Build v2" is a good idea, we seem to need more requirements to have a more concrete decisions on the `conceptual integrity` we want to have for this repository. My only 2 cents on this is that I would aim to have common ideas reflected later on the code, rather than very good ideas that doesn't fit between each other. I fear not sticking to what the `integrity` of the Build, BuildRun, etc is, will later be costly to this project.

One more question, when you say:
> We should definitely fork a separate discussion on what we should support and what we shouldn't.

Is this separate discussion dependant on the outcome of "Building CLI/UI tooling and experiences around Build v2" ?

Reply:
> My only 2 cents on this is that I would aim to have common ideas reflected later on the code, rather than very good ideas that doesn't fit between each other

Absolutely.

> I fear not sticking to what the integrity of the Build, BuildRun, etc is, will later be costly to this project.

Agreed.
In fact your thoughts on comment 1 and comment 2 also resonate with what I had in mind.
Let me know if you have any concerns with the discussion so far!


> Is this separate discussion dependent on the outcome of "Building CLI/UI tooling and experiences around Build v2" ?

No, this is an independent discussion. I have half-baked questions in my mind around 
- Should build v2 project worry about build triggers at all? 
- Wouldn't people anyway always use their own CI like travis or tekton :)
- Should we focus only on an amazing 'Build Image' experience.